File: libs/ocaml/trinity-vsa/lib/trinity_vsa.ml
(** Trinity VSA - Vector Symbolic Architecture with balanced ternary arithmetic *)

type trit = int (** -1, 0, or +1 *)
type trit_vector = int array

let zeros dim = Array.make dim 0

let random dim seed =
  Random.init seed;
  Array.init dim (fun _ -> Random.int 3 - 1)

let bind a b =
  if Array.length a <> Array.length b then
    failwith "Dimension mismatch";
  Array.map2 ( * ) a b

let unbind a b = bind a b

let bundle vectors =
  match vectors with
  | [] -> failwith "Empty vector list"
  | v :: _ ->
    let dim = Array.length v in
    Array.init dim (fun i ->
      let sum = List.fold_left (fun acc v -> acc + v.(i)) 0 vectors in
      if sum > 0 then 1 else if sum < 0 then -1 else 0)

let permute v shift =
  let dim = Array.length v in
  Array.init dim (fun i ->
    let idx = (i - shift + dim) mod dim in
    v.(idx))

let dot a b =
  if Array.length a <> Array.length b then
    failwith "Dimension mismatch";
  Array.fold_left2 (fun acc x y -> acc + x * y) 0 a b

let similarity a b =
  let d = float_of_int (dot a b) in
  let norm_a = sqrt (float_of_int (dot a a)) in
  let norm_b = sqrt (float_of_int (dot b b)) in
  if norm_a = 0. || norm_b = 0. then 0.
  else d /. (norm_a *. norm_b)

let hamming_distance a b =
  if Array.length a <> Array.length b then
    failwith "Dimension mismatch";
  Array.fold_left2 (fun acc x y -> if x <> y then acc + 1 else acc) 0 a b

File: libs/haskell/trinity-vsa/src/TrinityVSA.hs
{-# LANGUAGE BangPatterns #-}
-- | Trinity VSA - Vector Symbolic Architecture with balanced ternary arithmetic
module TrinityVSA
  ( Trit(..)
  , TritVector
  , zeros
  , random
  , bind
  , unbind
  , bundle
  , permute
  , dot
  , similarity
  , hammingDistance
  ) where

import qualified Data.Vector.Unboxed as V
import System.Random (mkStdGen, randomRs)

-- | Balanced ternary value: -1, 0, or +1
type Trit = Int

-- | Dense vector of trits
type TritVector = V.Vector Trit

-- | Create zero vector
zeros :: Int -> TritVector
zeros dim = V.replicate dim 0

-- | Create random vector
random :: Int -> Int -> TritVector
random dim seed = V.fromList $ take dim $ map (\x -> x `mod` 3 - 1) $ randomRs (0, 2) (mkStdGen seed)

-- | Bind two vectors (element-wise multiplication)
bind :: TritVector -> TritVector -> TritVector
bind = V.zipWith (*)

-- | Unbind (inverse of bind)
unbind :: TritVector -> TritVector -> TritVector
unbind = bind

-- | Bundle vectors via majority voting
bundle :: [TritVector] -> TritVector
bundle [] = error "Empty vector list"
bundle vs = V.generate dim $ \i ->
  let s = sum [v V.! i | v <- vs]
  in if s > 0 then 1 else if s < 0 then -1 else 0
  where dim = V.length (head vs)

-- | Circular permutation
permute :: TritVector -> Int -> TritVector
permute v shift = V.generate dim $ \i ->
  v V.! ((i - shift) `mod` dim)
  where dim = V.length v

-- | Dot product
dot :: TritVector -> TritVector -> Int
dot a b = V.sum $ V.zipWith (*) a b

-- | Cosine similarity
similarity :: TritVector -> TritVector -> Double
similarity a b
  | normA == 0 || normB == 0 = 0
  | otherwise = fromIntegral (dot a b) / (normA * normB)
  where
    normA = sqrt $ fromIntegral $ dot a a
    normB = sqrt $ fromIntegral $ dot b b

-- | Hamming distance
hammingDistance :: TritVector -> TritVector -> Int
hammingDistance a b = V.length $ V.filter id $ V.zipWith (/=) a b
